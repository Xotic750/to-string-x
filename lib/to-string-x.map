{"version":3,"sources":["lib/to-string-x.js"],"names":["f","exports","module","define","amd","g","window","global","self","this","returnExports","e","t","n","r","s","o","u","a","require","i","Error","code","l","call","length","1","_dereq_","isSymbol","ToString","value","TypeError","String","is-symbol","2","toStr","Object","prototype","toString","Symbol","symToStr","symStringRegex","isSymbolObject","valueOf","test"],"mappings":"CAAA,SAAUA,GAAG,GAAoB,gBAAVC,UAAoC,mBAATC,QAAsBA,OAAOD,QAAQD,QAAS,IAAmB,kBAATG,SAAqBA,OAAOC,IAAKD,UAAUH,OAAO,CAAC,GAAIK,EAAkCA,GAAb,mBAATC,QAAwBA,OAA+B,mBAATC,QAAwBA,OAA6B,mBAAPC,MAAsBA,KAAYC,KAAKJ,EAAEK,cAAgBV,MAAO,WAAqC,MAAO,SAAUW,GAAEC,EAAEC,EAAEC,GAAG,QAASC,GAAEC,EAAEC,GAAG,IAAIJ,EAAEG,GAAG,CAAC,IAAIJ,EAAEI,GAAG,CAAC,GAAIE,GAAkB,kBAATC,UAAqBA,OAAQ,KAAIF,GAAGC,EAAE,MAAOA,GAAEF,GAAE,EAAI,IAAGI,EAAE,MAAOA,GAAEJ,GAAE,EAAI,IAAIhB,GAAE,GAAIqB,OAAM,uBAAuBL,EAAE,IAAK,MAAMhB,GAAEsB,KAAK,mBAAmBtB,EAAE,GAAIuB,GAAEV,EAAEG,IAAIf,WAAYW,GAAEI,GAAG,GAAGQ,KAAKD,EAAEtB,QAAQ,SAASU,GAAG,GAAIE,GAAED,EAAEI,GAAG,GAAGL,EAAG,OAAOI,GAAEF,EAAEA,EAAEF,IAAIY,EAAEA,EAAEtB,QAAQU,EAAEC,EAAEC,EAAEC,GAAG,MAAOD,GAAEG,GAAGf,QAAkD,IAAI,GAA1CmB,GAAkB,kBAATD,UAAqBA,QAAgBH,EAAE,EAAEA,EAAEF,EAAEW,OAAOT,IAAID,EAAED,EAAEE,GAAI,OAAOD,KAAKW,GAAG,SAASC,EAAQzB,EAAOD,IAwC10B,WAEA,YAEA,IAAI2B,GAAWD,EAAQ,YAoBvBzB,GAAOD,QAAU,QAAS4B,UAASC,GACjC,GAAIF,EAASE,GACX,KAAM,IAAIC,WAAU,4CAEtB,OAAOC,QAAOF,SAIfG,YAAY,IAAIC,GAAG,SAASP,EAAQzB,EAAOD,GAC9C,YAEA,IAAIkC,GAAQC,OAAOC,UAAUC,QAG7B,IAFmC,kBAAXC,SAA6C,gBAAbA,UAExC,CACf,GAAIC,GAAWD,OAAOF,UAAUC,SAC5BG,EAAiB,iBACjBC,EAAiB,QAASA,gBAAeZ,GAC5C,MAA+B,gBAApBA,GAAMa,WACVF,EAAeG,KAAKJ,EAAShB,KAAKM,IAE1C5B,GAAOD,QAAU,QAAS2B,UAASE,GAClC,GAAqB,gBAAVA,GAAsB,OAAO,CACxC,IAA0B,oBAAtBK,EAAMX,KAAKM,GAAgC,OAAO,CACtD,KACC,MAAOY,GAAeZ,GACrB,MAAOnB,GACR,OAAO,QAITT,GAAOD,QAAU,QAAS2B,UAASE,GAElC,OAAO,aAIE,IAAI","file":"lib/to-string-x.min.js"}