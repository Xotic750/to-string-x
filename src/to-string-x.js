/**
 * @file ES6-compliant shim for ToString.
 * @see {@link http://www.ecma-international.org/ecma-262/6.0/#sec-tostring|7.1.12 ToString ( argument )}
 * @version 2.0.0
 * @author Xotic750 <Xotic750@gmail.com>
 * @copyright  Xotic750
 * @license {@link <https://opensource.org/licenses/MIT> MIT}
 * @module to-string-x
 */

import isSymbol from 'is-symbol';

/**
 * The abstract operation ToString converts argument to a value of type String.
 *
 * @param {*} [value] - The value to convert to a string.
 * @throws {TypeError} If `value` is a Symbol.
 * @returns {string} The converted value.
 */
export default function ToString(value) {
  if (isSymbol(value)) {
    throw new TypeError('Cannot convert a Symbol value to a string');
  }

  return ''.constructor(value);
}
