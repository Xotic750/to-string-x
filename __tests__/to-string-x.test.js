import noop from 'lodash/noop';
import $toString from 'src/to-string-x';

const hasSymbol = typeof Symbol === 'function' && typeof Symbol('') === 'symbol';
const ifSymbolIt = hasSymbol ? it : xit;

describe('basic tests', function() {
  it('should return a string for everything', function() {
    expect.assertions(1);
    const values = [true, 'abc', 1, null, undefined, noop, [], /r/];

    const expected = values.map(String);
    const actual = values.map($toString);
    expect(actual).toStrictEqual(expected);
  });

  it('should throw for Object.create(null)', function() {
    expect.assertions(1);
    expect(function() {
      $toString(Object.create(null));
    }).toThrowErrorMatchingSnapshot();
  });

  ifSymbolIt('should throw for Symbol', function() {
    expect.assertions(2);

    const sym = Symbol('foo');
    expect(function() {
      $toString(sym);
    }).toThrowErrorMatchingSnapshot();

    const symObj = Object(sym);
    expect(function() {
      $toString(Object(symObj));
    }).toThrowErrorMatchingSnapshot();
  });
});
