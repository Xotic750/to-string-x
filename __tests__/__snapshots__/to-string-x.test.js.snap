// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic tests should throw for Object.create(null) 1`] = `"Cannot convert object to primitive value"`;

exports[`basic tests should throw for Symbol 1`] = `"Cannot convert a Symbol value to a string"`;

exports[`basic tests should throw for Symbol 2`] = `"Cannot convert a Symbol value to a string"`;
